### PR TITLE
Draft: Add support for ICMPv4 data part

### LINKED ISSUE #46: Add support for ICMPv4 data part

### PR DESCRIPTION
**Pull request by hgupta12**
This PR adds ICMPv4 data dissection.
Fixes #46

### CODE REVIEW

File: src/layers/icmp.rs [hunk 1]
```diff
@@ -134,6 +197,36 @@ impl Layer for ICMP {
                 })
             }
         };
+
+
+        self.data = match self.icmp_type {
+            ICMP_ECHO_REPLY | ICMP_ECHO_REQUEST => {
+                decoded = bytes.len();
+                IcmpData::Unknown(IcmpUnknownData { data: bytes[8..].to_vec() })
```
Comment (gabhijit) [inline src/layers/icmp.rs:205]: This is more of a question. Can you please point to the relevant section in the RFC for me to better understand what you are trying to do?

**Review by gabhijit — COMMENTED** [review]
Codewise this looks okay. If you can point me to relevant section in the RFC, I will try to read a bit and see if the organization of Data makes sense or is there some better (more semantic) way of handling that.

File: src/layers/icmp.rs (hunk 1 above)
Comment (hgupta12) [reply src/layers/icmp.rs:205]: image
Here, the data part in ICMP Echo Request and Reply are unknown and variable. The reply just sends back the same data it gets in the request.

Comment (gabhijit) [reply src/layers/icmp.rs:205]: Better to call it just `Data` than `UnknownData`. That sounds a bit confusing.

File: src/layers/icmp.rs [hunk 2]
```diff
@@ -59,6 +66,42 @@ pub struct IcmpUnsupported {
     unsupported: Vec<u8>,
 }
 
+#[derive(Default, Debug, Serialize)]
+#[serde(untagged)]
+pub enum IcmpData {
+    #[default]
+    None,
+    Unknown(IcmpUnknownData),
```
Comment (gabhijit) [inline src/layers/icmp.rs:74]: nitpick: Probably better to call the Variants as follows - 

`Data(Data)`, `TimeStamp(TimestampData)`, `Error(OriginalDatagram)`

**Review by gabhijit — COMMENTED** [review]
I am thinking something along the following lines as a change - 

Basically - what you are trying to do during the dissection is - you are abstracting out 'sections' of the header and then remaining part is 'data' depending upon the message type. 

The `rest_of_header` and `data` fields can be merged into single `IcmpType` part. 

For instance right now - the Timestamp Packet when displayed will be displayed as `Echo` packet with `Timestamp` data - which doesn't quite look good. Instead it should simply be called `Timestamp` packet (with data included as vec in the corresponding `IcmpType` variant).

Let me know if that makes sense.

**Comment by hgupta12** [comment]
@gabhijit sir, I have merged the IcmpData enum into the IcmpType. This does look better than my original approach. Thanks for the input!

**Comment by gabhijit** [comment]
This is now merged in master! Thank you for your contribution!

**Pull request closed without merging**